Don't fallback to lkl mmap

SeanTAllen opened the pull request:
the fallback from enclave mmap to lkl mmap is a source of possible subtle bugs and is being removed. this is part 1 of a few incremental changes to mmap usage in order to address layering issues and bugs.

Commit:
Don't fallback to lkl mmap

SeanTAllen commented:
@prp @davidchisnall CI is running without the fallback to lkl mmap.

Commit:
Allow SHARED mmapings

Commit:
Baby steps

Commit:
Add missing newlines to end of json files

Commit:
Don't map more of the file than requested

Previously it was mmapping length but would read the entire file in.
From my reading of the man page that is wrong.

Commit:
Remove MAP_FAILED from enclave_mmap

Commit:
Don't move the file offset when mmapping a file

Commit:
Fix that newlines at end thing

Commit:
Clean up mmap files flags supported code

Commit:
Fix incorrect logic in MAP_FAILED removal

Commit:
Remove incorrect mmap return value usage

Commit:
README touch

Commit:
EACCES instead of EBADF

Either might be the error. LTP tests expect EACCES. We'd need
to check errno to get it "correct". We are trying to remove
errno access at this layer.

Commit:
Return correct mmap file read failure

Commit:
Undo touch of README

Commit:
Update enclave config reference files

Commit:
Momentary change:

Commit:
Is this our jvm problem?

Commit:
Fix a dur

Commit:
Move some code over the mem overrides

Commit:
Always return -EBADF

Commit:
Call pread directly

I'm getting an unbalanced put otherwise. That override with pread_fn was
probably bad as I only did half the work that David had on it.

SeanTAllen commented on src/lkl/posix-host.c:
```diff
@@ -681,6 +681,13 @@ static void *host_malloc(size_t size)
 	{
 		SGXLKL_ASSERT(kernel_mem == NULL);
 		kernel_mem = enclave_mmap(0, size, 0, PROT_READ | PROT_WRITE, 0);
+        if (((intptr_t)kernel_mem) < 0)
```
the formatting is off here due to #665. I am planning on addressing after this PR. I didn't want to muddy up the diff with a ton of formatting.

SeanTAllen commented on src/lkl/syscall-overrides-mem.c:
```diff
@@ -0,0 +1,96 @@
+            // Read file into memory
+            size_t readb = 0;
+            ssize_t ret = 0;
+            while ((ret = pread(
```
if I use pread_fn I get an unbalanced lkl_cpu_put error from the java hello world error.

I wasn't expecting that @davidchisnall and would like to know what faulty assumption I made about this half-way point to the changes in #604. 

I can move things back out of the syscall-overrides-mem etc as without using pread_fn, all this can go back into enclave_mem.c

davidchisnall replied:
That's odd.  This should be `lkl_sys_pread`, not the libc wrapper, but I don't understand why calling a syscall implementation function would cause problems but making `lkl_syscall` reentrant (which is *really* wasn't designed to be) would work.

SeanTAllen replied:
updated.

Commit:
Remove syscall-overrides work

As I'm not using pread_fn, it no longer makes sense as part of this PR
and would be part of the finishing of 604 once that is unblocked.

Commit:
Clean up diff

SeanTAllen commented:
Because CI round-trip integration is broken with GH at the moment, passing CI is here: https://dev.azure.com/sgx-lkl/sgx-lkl/_build/results?buildId=1547&view=results

SeanTAllen commented:
ill need to rebase sgx-lkl-musl before this is merged. that can be part of the bumping.

all the same code is on both, but there was an issue with it not pointing to a squashed very of some commits when I branched.

mikbras commented on src/enclave/enclave_mem.c:
```diff
@@ -97,59 +98,59 @@ static size_t addr_to_index(void* addr)
-            lseek(fd, offset, SEEK_SET);
-            size_t readb = 0, ret;
-            while ((ret = read(fd, ((char*)mem) + readb, length - readb)) > 0)
+            size_t readb = 0;
```
Just wondering whether readb was supposed contain something non-zero. Otherwise, I'm not sure what its purposes is.

SeanTAllen replied:
It's how many bytes your have gotten back from pread.

in case pread returns less than legnth and you need to read again.

mikbras reviewed (approved):
Sean, looks great. I couldn't find any problems. Thank you for doing this work.

SeanTAllen commented:
Ok because of reverts to sgx-lkl-musl. This PR looks to be messed up

Commit:
Bump sgx-lkl-musl

Commit:
Add commits lost during rebase.

Commit:
Add more changes lost during final rebase
